### PR TITLE
Init

### PR DESCRIPTION
**Pull request by bludot**

### CODE REVIEW

**Comment by bludot** [comment]
Some things still need to be done. But the gist is there.

TODO:

* proper es modules support (for example: 'ui-framework/components/Button')
* Clean up packages needed to install (right now I think im installing more than I need to)

File: examples/main/package.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (Vincent-Alexandrine) [inline examples/main/package.json]: Well, doesn't seem to be our.

File: examples/main/src/index.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
```
Comment (Vincent-Alexandrine) [inline examples/main/src/index.html]: Parcel ?

File: examples/main/src/index.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+);
+
+/*
+const App = () => (
+  <h1>Hello</h1>
+);
+*/
+console.log(Button);
+ReactDOM.render(
+  <ThemeProvider theme={themes.selfapyTheme}>
```
Comment (Vincent-Alexandrine) [inline examples/main/src/index.js]: `themes.selfapyTheme`
wasn't like this before. Place it back

File: src/components/Button/Button.jsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import React, { PureComponent } from 'react';
+import PropTypes from 'prop-types';
+
+import { variants, types } from './constants';
+import { Container } from './style';
+
+export default class Button extends PureComponent {
```
Comment (Vincent-Alexandrine) [inline src/components/Button/Button.jsx]: Shouldn't be a PureComponent, but i'm gonna take care of it myself

File: src/components/Button/index.jsx [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export { Container as StyledContainer } from './style';
```
Comment (Vincent-Alexandrine) [inline src/components/Button/index.jsx]: No jsx inside, should stay .js

File: src/components/Test/Test.jsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import React, { PureComponent } from 'react';
```
Comment (Vincent-Alexandrine) [inline src/components/Test/Test.jsx]: remove the `components/Test` folder

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    files[r2.match(/\/(.+?)(\/.*|\/).*.jsx/)[1]] = context(r2).default || context(r2);
+  });
+}
+
+importAll(require.context('./components/', true, /(.+?)\/\1\.jsx$/));
+
+module.exports = {
+  ...files,
+  themes,
+  ThemeProvider,
```
Comment (Vincent-Alexandrine) [inline src/index.js]: don't export the ThemeProvider among the library

File: styleguide/colors/index.js [hunk 8]
```diff
@@ -0,0 +1,12 @@
+export default {
```
Comment (Vincent-Alexandrine) [inline styleguide/colors/index.js]: Shouldn't be there

File: styleguide/theme.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import colors from './colors';
```
Comment (Vincent-Alexandrine) [inline styleguide/theme.js]: shouldn't be there

File: webpack.config.js [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+  entry: './src/index.js',
+  output: {
+    path: path.resolve(__dirname, 'lib'),
+    filename: 'library.js',
+    libraryTarget: 'umd',
+    umdNamedDefine: true,
+  },
+  resolve: {
+    alias: {
+      react: path.resolve(__dirname, './node_modules/react'),
```
Comment (Vincent-Alexandrine) [inline webpack.config.js]: alias 'src' ?

File: webpack.config.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+  },
+  devtool: 'none',
+  node: {
+    net: 'empty',
+    tls: 'empty',
+    dns: 'empty',
+  },
+  resolve: {
+		alias: {
+			'src': path.join(__dirname, './', 'src'),
```
Comment (Vincent-Alexandrine) [inline webpack.config.js]: Ah ... there
well, "resolve" keys has to be merged

**Comment by Sergeytroinin** [comment]
I'd keep it just a component lib without any example app to render them. We have styleguidist, where we can use `ThemeProvider` and show all the components, that we have. So few points:
- It makes sense to keep the same structure with separate folder for each component and one file with component itself. The we can have one index file to reexport it from lib
```
components/
    MyComponent/
        MyComponent.js
        MyComponent.test.js
        README.md // we need it for styleguidist
        index.js
index.js // everything can be reexported here
```
- To keep this structure we can use the same `redux-cli` and blueprints from boilerplate.
- `ThemeProvider` can be added to styleguidist through `Wrapper`. Look [here](https://github.com/styleguidist/react-styleguidist/blob/master/docs/Thirdparties.md#adding-styled-components-themeprovider) for details
- We already have a lot of this implemented in the boilerplate/generator/config repos. So feel free to reuse it. I suppose you can even create new project with generator and remove unused parts for redux/screens, etc. But it's up to you.
- I know, that it's not final version, but would be great to have some README and CHANGELOG files. Changelog will allow us to track all the changes and keep proper versioning for this library. I'd recommend use such structure https://keepachangelog.com/en/1.0.0/. We already do it for config and generator.

File: examples/main/src/index.html (hunk 2 above)
Comment (bludot) [reply examples/main/src/index.html]: should be there anyways. Webpack adds it too.

File: examples/main/src/index.js (hunk 3 above)
Comment (bludot) [reply examples/main/src/index.js]: what was it before?

**Comment by bludot** [comment]
`index.js // everything can be reexported here` for this part here, I'm leveraging webapp dynamic imports to generate this file so we don't have to populate it with all the components every time we add a new one. 

I also will try to use another package to generate an index file when we generate an es build as this allows us to do `import Button from 'ui-framework/components/Button';` which will help because we dont want to import the whole lib, especially if we are going to do code splitting. I am thinking of using rollup for this as it does support bundling es modules. See univ-router which uses rollup to bundle both es and non es modules.

I did not want to use redux-cli because I want to keep redux completely out of this. IMO, the component library should have the bare minimum and I dont see redux playing a part in them. Im pretty sure a component should not have its own redux store. If it turns out we might use it here, then we can add later. It should not be hard to create 5 files (component, style, test, readme, index).

**Comment by bludot** [comment]
Not needed

**Pull request closed without merging**